src/BenchmarksApps/TcpEcho/SocketDuplexPipeConnection.cs

```diff
@@ -0,0 +1,92 @@
+﻿using System;
+using System.IO.Pipelines;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace TcpEcho
+{
+    public class SocketDuplexPipeConnection : IDuplexPipe, IThreadPoolWorkItem, IAsyncDisposable
+    {
+        public PipeReader Input { get; }
+        public PipeWriter Output { get; }
+
+        private readonly Task _socketTasks;
+
+        public SocketDuplexPipeConnection(Socket socket)
+        {
+            var input = new Pipe();
+            var output = new Pipe();
+
+            Input = input.Reader;
+            Output = output.Writer;
+
+            var readTask = ReadFromSocket();
+            var writeTask = WriteToSocket();
+
+            _socketTasks = Task.WhenAll(readTask, writeTask);
+
+            async Task ReadFromSocket()
+            {
+                while (true)
+                {
+                    var buffer = input.Writer.GetMemory(2048);
+                    var read = await socket.ReceiveAsync(buffer, SocketFlags.None);
+
+                    if (read == 0)
+                    {
+                        break;
+                    }
+
+                    await input.Writer.FlushAsync();
```

Comment:
same here, use `read`?

Reply:
Good catch. Given how prescriptive the default Pipe implementation can be fore some things, it's surprising that it's OK with a missing call to PipeWriter.Advance(). I'm guessing nothing is actually being written to the socket.

Reply:
OOps